File: cpp-linux/annotations/annotation_replies/annotation_replies.cpp
// annotation_replies -- C++ port of examples\Vb6\annotations\annotation_replies\annotation_replies.bas
// A square annotation with a reply, and a reply to that reply.
#include <lumaspdf.h>
#include <cstdio>
#include <string>


static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    return 0;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, PDFError);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);

    pdfAppend(pdf);
    // To see the reply click on the annotation
    SI32 annot = pdfSquareAnnotA(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, csDeviceRGB, "Jim", "Test", "Just test...");
    SI32 reply = pdfSetAnnotMigrationStateA(pdf, annot, asCreateReply, "Harry");
    pdfSetAnnotStringA(pdf, reply, asContent, "This is a reply!");

    reply = pdfSetAnnotMigrationStateA(pdf, reply, asCreateReply, "Jim");
    pdfSetAnnotStringA(pdf, reply, asContent, "This is a reply to a reply!");
    pdfEndPage(pdf);

    if(pdfHaveOpenDoc(pdf) != 0){
        std::string outFile = exeDir(argv[0]) + "/out.pdf";
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){
            pdfDeletePDF(pdf);
            return 0;
        }
        if(pdfCloseFile(pdf) != 0){
            printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
        }
    }

    pdfDeletePDF(pdf);
    return 0;
}

File: vbnet/annotations/annotation_replies/annotation_replies.vb
' annotation_replies -- VB.NET port of examples\Vb6\annotations\annotation_replies
Imports System
Imports System.IO
Imports LumasPdfSdk

Module AnnotationReplies
    Private Const NO_COLOR As UInteger = &HFFFFFFF1UI
    Private errDel As TErrorProc

    Public Function PDFError(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Return 0
    End Function

    Sub Main()
        Dim annot As Integer, reply As Integer
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        errDel = New TErrorProc(AddressOf PDFError)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)
        LumasPdf.pdfCreateNewPDFW(pdf, "")

        LumasPdf.pdfSetPageCoords(pdf, TPageCoord.pcTopDown)

        LumasPdf.pdfAppend(pdf)
        annot = LumasPdf.pdfSquareAnnotW(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, TPDFColorSpace.csDeviceRGB, "Jim", "Test", "Just test...")
        reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, annot, TAnnotState.asCreateReply, "Harry")
        LumasPdf.pdfSetAnnotStringW(pdf, reply, TAnnotString.asContent, "This is a reply!")

        reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, reply, TAnnotState.asCreateReply, "Jim")
        LumasPdf.pdfSetAnnotStringW(pdf, reply, TAnnotString.asContent, "This is a reply to a reply!")
        LumasPdf.pdfEndPage(pdf)

        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            Dim outFile As String = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf")
            If LumasPdf.pdfOpenOutputFileW(pdf, outFile) = 0 Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            If LumasPdf.pdfCloseFile(pdf) <> 0 Then
                Console.WriteLine("PDF file """ & outFile & """ successfully created!")
            End If
        End If

        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module
